File: 208_Implement_Trie-Prefix_Tree/208_1.cpp
#include <iostream>
#include <string>
#include <vector>
#include <climits>
#include <algorithm>
#include <cstring>
using namespace std;

class Trie {
private:
    bool isEnd;
    Trie *next[26];
public:
    Trie() {
        isEnd = false;
        memset(next, 0, sizeof(next));
    }

    void insert(string word) {
        Trie *node = this;
        for (char c : word) {
            if (node->next[c - 'a'] == nullptr)
                node->next[c - 'a'] = new Trie();
            node = node->next[c - 'a'];
        }
        node->isEnd = true;
    }

    bool search(string word) {
        Trie *node = this;
        for (char c : word) {
            node = node->next[c - 'a'];
            if (node == nullptr)
                return false;
        }
        return node->isEnd;
    }

    bool startsWith(string prefix) {
        Trie *node = this;
        for (char c : prefix) {
            node = node->next[c - 'a'];
            if (node == nullptr)
                return false;
        }
        return true;
    }
};


int main() {
    Trie *trie = new Trie();
    trie->insert("apple");
    cout << boolalpha;
    cout << trie->search("apple") << endl;
    cout << trie->startsWith("app") << endl;
}

File: 208_Implement_Trie-Prefix_Tree/208_1.py
class Trie:
    def __init__(self):
        self.next_ = [None for _ in range(26)]
        self.is_end = False

    def insert(self, word: str) -> None:
        node = self
        for w in word:
            i = ord(w) - ord('a')
            if not node.next_[i]:
                node.next_[i] = Trie()
            node = node.next_[i]
        node.is_end = True

    def search(self, word: str) -> bool:
        node = self
        for w in word:
            i = ord(w) - ord('a')
            if not node.next_[i]:
                return False
            node = node.next_[i]
        return node.is_end

    def startsWith(self, prefix: str) -> bool:
        node = self
        for w in prefix:
            i = ord(w) - ord('a')
            if not node.next_[i]:
                return False
            node = node.next_[i]
        return True
